README.md
APP-SPEECH-BackTalk
===================

Free speech system for OS/2


LICENSE
===============
* 

COMPILE TOOLS
===============
* 

AUTHORS
===============
* Samuel Audet
* Jim Little

LINKS
===============
* 


history.txt
BackTalk project history
~~~~~~~~~~~~~~~~~~~~~~~~
2.01 - Removed MCI_DOS_QUEUE Warp 4 specific MCI_OPEN flag (sorry).  Now
       should work with Warp 3 no problem.  And one less useless thread.
     - Fixed possible hang.

2.0  - Complete remake of BackTalk main engine.
       It is now integral part of the audio engine (which has also been
       reworked a lot).  You will notice major performance, solidity and
       functionality increase.  Noticeably amplification, realtime volume
       adjusting, and realtime queue status watching.  To access BackTalk,
       there are now ready-to-use functions in DLL for both REXX scripts
       and programming languages such as C.  Enjoy!

1.3  - Included PMAUD 1.1 as a part of a kludge.
     - Installation now proposes testing of 7 options (including kludge ;)
     - SPKIRC, speaking IRC REXX script!  Many options
     - SPCHTIME -> SPKCLOCK
     - SPCHTMR  -> SPKTIMER
     - Whole new SPKCLOCK from J Hulley-Miller <jhm@pobox.com>
     - Enhanced and bug fixed SPKTIMER
     - SPEECHQ.CMD will now execute SPKSTART.CMD after loading
     - SPKMAIL now specifies high priority and fowarded e-mail
     - Included Info-ZIP UNZIP.EXE (easier install?)

1.2  - Included Rsynth 2.1 and 2.2.
     - New speaking REXX scripts: SPCHTIME and SPCHTMR.
     - Readme updated accordingly, plus a "how to use in REXX" section.
     - Removed 16 bit playback by default.
     - Fixed bugs here and there

1.1  Revised by Samuel Audet
     - Bug fixed and ameliorated SPKMAIL and SPEECHQ.
     - Included Rsynth 2.2.
     - Changed the INSTALL program so I could understand it.
     - Updated readme accordingly.

1.0  Original release by Jim Little


uptime/dqps.h
#ifndef DQPS_H
#define DQPS_H
/*

  The explanation for all the structures are in Kai Uwe Rommel's
  documentation available in
  
  at ftp.leo.org in directory /pub/comp/os/os2/doc as proc.zip
  
*/



/*
   
   declare the 16-bit DosQProcStatus in such a manner that gcc-emx will accept
   it.
   
   
*/

unsigned _THUNK_PASCAL_FUNCTION (DosQProcStatus) (ULONG* buffer, USHORT buffer_size); 

typedef struct qsGrec_s {
   ULONG cThrds;                /* number of threads in use */
   ULONG Reserved1;
   ULONG Reserved2;
} qsGrec_t;

typedef struct qsTrec_s {
   ULONG RecType;               /* 00: Record Type */
                                /*     Thread rectype = 0x00000100 */
   USHORT tid;                  /* 04: thread ID */
   USHORT slot;                 /* 06: "unique" thread slot number */
   ULONG sleepid;               /* 08: sleep id thread is sleeping on */
   ULONG priority;              /* 0C: thread priority */
   ULONG systime;               /* 10: thread system time */
   ULONG usertime;              /* 14: thread user time */
   UCHAR state;                 /* 18: thread state */
   UCHAR padchar ;              /* 19: */
   USHORT padshort ;            /* 1A: */
} qsTrec_t;                     /* Total size: 28 bytes */

typedef struct qsPrec_s {
   ULONG RecType;               /* 00: type of record being processed */
                                /*     process rectype = 0x00000001   */
   qsTrec_t * _Seg16 pThrdRec;  /* 04: ptr to 1st thread rec for this prc */
   USHORT pid;                  /* 08: process ID */
   USHORT ppid;                 /* 0A: parent process ID */
   ULONG type;                  /* 0C: process type */
   ULONG stat;                  /* 10: process status */
   ULONG sgid;                  /* 14: process screen group */
   USHORT hMte;                 /* 18: program module handle for process */
   USHORT cTCB;                 /* 1A: # of TCBs in use in process */
   ULONG Reserved1 ;
   ULONG Reserved2 ;
   USHORT c16Sem;               /* 24: # of 16 bit system sems in use by proc */
   USHORT cLib;                 /* 26: number of runtime linked libraries */
   USHORT cShrMem;              /* 28: number of shared memory handles */
   USHORT Reserved3 ;
   USHORT * _Seg16 p16SemRec;   /* 2C: ptr to head of 16 bit sem inf for proc */
   USHORT * _Seg16 pLibRec;     /* 30: ptr to list of runtime lib in use by process */
   USHORT * _Seg16 pShrMemRec;  /* 34: ptr to list of shared mem handles in use by process */
   ULONG Reserved4 ;            /* 38: */
} qsPrec_t;                     /* Total size: 60 bytes */

typedef struct qsS16Headrec_s { 
   ULONG SRecType;              /* semaphore rectype = 3 */
   ULONG Reserved1;             /* overlays NextRec of 1st qsS16rec_t*/
   ULONG Reserved2;
   ULONG S16TblOff;             /* index of first semaphore,SEE PSTAT OUTPUT*/
                                /* System Semaphore Information Section     */
} qsS16Headrec_t;
 
typedef struct qsS16rec_s { 
   ULONG NextRec;               /* offset to next record in buffer*/
   UINT  s_SysSemOwner ;        /* thread owning this semaphore   */
   UCHAR s_SysSemFlag ;         /* system semaphore flag bit field*/
   UCHAR s_SysSemRefCnt ;       /* number of references to this   */
                                /*   system semaphore             */
   UCHAR s_SysSemProcCnt ;      /*number of requests by sem owner*/
   UCHAR Reserved1;
   ULONG Reserved2;
   UINT  Reserved3;
   CHAR  SemName;               /* start of semaphore name string */
} qsS16rec_t;
 
 
typedef struct qsMrec_s { 
   ULONG  MemNextRec;           /* offset to next record in buffer */
   USHORT hmem;                 /* handle for shared memory */
   USHORT sel;                  /* shared memory selector */
   USHORT refcnt;               /* reference count */
   CHAR   Memname;              /* start of shared memory name string */
} qsMrec_t;
 
typedef struct qsLrec_s { 
   void * _Seg16 pNextRec;      /* pointer to next record in buffer */
   USHORT hmte;                 /* handle for this mte */
   USHORT Reserved1;            /* Reserved */
   ULONG  ctImpMod;             /* # of imported modules in table */
   ULONG  Reserved2;            /* Reserved */
   ULONG  Reserved3;            /* Reserved */
   UCHAR * _Seg16 pName;        /* ptr to name string following stru*/
} qsLrec_t;
 
typedef struct qsPtrRec_s {
   qsGrec_t *pGlobalRec;        /* ptr to the global data section */
   qsPrec_t *pProcRec;          /* ptr to process record section  */
   qsS16Headrec_t *p16SemRec;   /* ptr to 16 bit sem section      */
   qsMrec_t *pShrMemRec;        /* ptr to shared mem section      */
   qsLrec_t *pLibRec;           /* ptr to exe module record section */
} qsPtrRec_t;

#endif


btalk/clients/source/spktimer.c
#define INCL_DOS
#include <os2.h>
#include <time.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "btclient.h"

void help(char *program);

int main(int argc, char *argv[])
{
   double time;
   char temp[256];
   int i;

   HEV sem;
   HTIMER timer;
   ULONG resetcount;

   int minsleft, secsleft, timersecs, urgsecsleft, countsecs;

   int timermins = 60;      /* countdown time, minutes  */
   int lap = 300;           /* lap between spoken time left, seconds */
   char *message = " ";     /* message pronounced after the time left */

   int urgminsleft = 20;    /* Falls in urgent mode after urgminleft time, minutes */
   int urglap = 180;        /* Urgent mode lap, seconds  */
   char *urgmsg = " ";      /* Urgent mode message  */

   char *alertmsg = " ";    /* Alert message when timer has stopped */

   for(i = 0; i < argc; i++)
      if ((argv[i][0] == '/') || (argv[i][0] == '-'))
         switch(toupper(argv[i][1]))
         {
            case 'T': if(argv[i][2] == '1')
                         timermins = atoi(argv[++i]);
                      else if(argv[i][2] == '2')
                         urgminsleft = atoi(argv[++i]);
                      break;
            case 'L': if(argv[i][2] == '1')
                         lap = atoi(argv[++i]);
                      else if(argv[i][2] == '2')
                         urglap = atoi(argv[++i]);
                      break;
            case 'M': if(argv[i][2] == '1')
                         message = argv[++i];
                      else if(argv[i][2] == '2')
                         urgmsg = argv[++i];
                      else if(argv[i][2] == '3')
                         alertmsg = argv[++i];
                      break;

            case '?': help(argv[0]); return 0;
         }

   timersecs = timermins * 60;      /* Converts minutes into seconds. */
   urgsecsleft = urgminsleft * 60;


   /* Set some abnormalities */

   if(lap > timersecs) lap = timersecs;
   if(urgsecsleft > timersecs) urgsecsleft = 0;
   if(urglap > urgsecsleft) urglap = urgsecsleft;
   if(urglap > lap) urglap = lap;

   /* Give some visual information */

   printf("countdown time: %d minutes.\n",timersecs/60);
   printf("laps time: %d seconds.\n",lap);
   printf("message: %s\n", message);
   printf("\n");
   printf("fall in urgent mode at %d minutes left.\n",urgsecsleft/60);
   printf("laps time in urgent mode: %d seconds.\n",urglap);
   printf("urgent message: %s\n", urgmsg);
   printf("\n");
   printf("alert message: %s\n",alertmsg);

   if(DosCreateEventSem(NULL,&sem,DC_SEM_SHARED,FALSE)) return 1;

   sprintf(temp,"Starting a %d minute countdown %s.",timersecs/60,message);
   BTSayString(temp,0,0,0);
   time = clock()/CLOCKS_PER_SEC;
   countsecs = timersecs;

   if(!DosStartTimer(lap * 1000,(HSEM) sem,&timer))
   {
      while(countsecs > urgsecsleft)
      {
         if(DosWaitEventSem(sem, -1)) return 1;
         DosResetEventSem(sem,&resetcount);
         countsecs = timersecs - (clock()/CLOCKS_PER_SEC - time);
         minsleft = countsecs / 60;
         secsleft = countsecs % 60;
         sprintf(temp,"There is %d minutes and %d seconds left %s.",minsleft,secsleft,message);
         BTSayString(temp,0,0,0);
         if(countsecs <= lap)
            while(countsecs > urgsecsleft)
            {
               DosSleep(1000);
               countsecs = timersecs - (clock()/CLOCKS_PER_SEC - time);
            }
      }
   DosStopTimer(timer);
   DosResetEventSem(sem,&resetcount);
   }

   if(!DosStartTimer(urglap * 1000,(HSEM) sem,&timer))
   {
      while(countsecs > 0)
      {
         if(DosWaitEventSem(sem, -1)) return 1;
         DosResetEventSem(sem,&resetcount);
         countsecs = timersecs - (clock()/CLOCKS_PER_SEC - time);
         minsleft = countsecs / 60;
         secsleft = countsecs % 60;
         sprintf(temp,"There is only %d minutes and %d seconds left %s.",minsleft,secsleft,message);
         BTSayString(temp,0,0,0);
         if(countsecs <= urglap)
            while(countsecs > 0)
            {
               DosSleep(1000);
               countsecs = timersecs - (clock()/CLOCKS_PER_SEC - time);
            }
      }
   DosStopTimer(timer);
   DosResetEventSem(sem,&resetcount);
   }

   BTSayString(alertmsg,0,0,0);

   return 0;

}

void help(char *program)
{
      printf("Speaking Timer - Part of the BackTalk project\n"
             "\n"
             "%s [-t1 <mins>] [-l1 <secs>] [-m1 <str>] [-t2 <mins>] [-l2 <secs>] [-m2 <str>] [-m3 <str>]\n"
             "\n"
             "-t1  countdown time (minutes)\n"
             "-l1  laps time (seconds)\n"
             "-m1  message\n"
             "\n"
             "-t2  time remaining to switch in urgent mode (minutes)\n"
             "-l2  laps time in urgent mode (seconds)\n"
             "-m2  message in urgent mode\n"
             "\n"
             "-m3  alert message when timer has stopped\n"
             "\n"
             "-?   Help screen\n"
             "\n"
             "ex.: [D:\\btalk] spktimer -t1 45 -t2 15 -m1 \"for your dentist rendezvous\" -m2 \n"
             "     \"for your VERY VERY important dentist rendezvous\" -l1 300 -l2 120 -m3 \n"
             "     \"Your dentist rendezvous!!\"\n",program);
}


btalk/clients/source/spkclock.c
#define INCL_DOS
#include <os2.h>
#include <time.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "btclient.h"

enum MODE { civil, offset, international } mode;

void help(char *program);

int main(int argc, char *argv[])
{
   time_t ltime;
   struct tm *stime;

   USHORT lapsecs = 0, lapmins = 15;
   BOOL hourstart = TRUE;

   int i;

   HEV sem;
   HTIMER timer;
   ULONG resetcount;

   char string[256],
        *temp = "international";

   mode = international;

   for(i = 0; i < argc; i++)
      if ((argv[i][0] == '/') || (argv[i][0] == '-'))
         switch(toupper(argv[i][1]))
         {
            case 'D': hourstart = FALSE; break;
            case 'L': lapmins = atoi(argv[++i]); break;
            case 'M':
               switch(toupper(argv[i][2]))
               {
                  case 'I': mode = international; temp = "international"; break;
                  case 'C': mode = civil; temp = "civil"; break;
                  case 'O': mode = offset; temp = "offset"; break;
               }
               break;
            case '?': help(argv[0]); return 0;
         }

   if(DosCreateEventSem(NULL,&sem,DC_SEM_SHARED,FALSE)) return 1;

   time(&ltime);
   stime = localtime(&ltime);
   printf("Started on %s",ctime(&ltime));
   printf("Running in %s mode.\n",temp);
   printf("Intervals between time announcements will be %u minutes.\n",lapmins);

   if(hourstart)
   {
      /* calculates number of second needed to fall even
         at nearest second on next lap */
      lapsecs = (lapmins - stime->tm_min % lapmins) * 60 - stime->tm_sec;
      printf("First Interval will be %u minutes and %u seconds.\n",lapsecs/60,lapsecs%60);
   }
   else
      lapsecs = lapmins*60;

   if(DosStartTimer(lapsecs * 1000,(HSEM) sem,&timer)) return 1;

   while(1)
   {
      strcpy(string,"It is ");

      if(mode == civil)
      {
         if(stime->tm_hour > 12)
            _itoa(stime->tm_hour - 12, strchr(string,'\0'),10);
         else if(stime->tm_hour == 0)
            strcat(string,"12"); /* midnight */
         else
            _itoa(stime->tm_hour, strchr(string,'\0'),10);

         strcat(string, " ");

         if(stime->tm_min < 10)
         {
            strcat(string,"O ");
            _itoa(stime->tm_min, strchr(string,'\0'),10);
         }
         else if(stime->tm_min != 0)
            _itoa(stime->tm_min, strchr(string,'\0'),10);

         if(stime->tm_hour > 11)
            strcat(string," P M");
         else
            strcat(string," AY M");

      }
      else if(mode == offset)
      {
         if(stime->tm_min != 0)
         {
            if(stime->tm_min == 15)
               strcat(string,"quarter past ");
            else if(stime->tm_min == 30)
               strcat(string,"half past ");
            else if(stime->tm_min == 45)
            {
               strcat(string,"quarter to ");
               stime->tm_hour++;
            }
            else if(stime->tm_min > 30)
            {
               _itoa(60 - stime->tm_min, strchr(string,'\0'),10);
               strcat(string," minutes to ");
               stime->tm_hour++;
            }
            else if(stime->tm_min < 30)
            {
               _itoa(stime->tm_min, strchr(string,'\0'),10);
               strcat(string," minutes past ");
            }
         }

         if(stime->tm_hour > 12)
            _itoa(stime->tm_hour - 12, strchr(string,'\0'),10);
         else if(stime->tm_hour == 0)
            strcat(string,"12"); /* midnight */
         else
            _itoa(stime->tm_hour, strchr(string,'\0'),10);

         if(stime->tm_min == 0) strcat(string," O'Clock");

      }
      else if(mode == international)
      {
         _itoa(stime->tm_hour, strchr(string,'\0'),10);
         strcat(string," hundred hours ");
         if(stime->tm_min > 0)
         {
            _itoa(stime->tm_min, strchr(string,'\0'),10);
            strcat(string," minutes");
         }
         if(stime->tm_sec > 0)
         {
            strcat(string," and ");
            _itoa(stime->tm_sec, strchr(string,'\0'),10);
            strcat(string," seconds");
         }
      }
      strcat(string,".");

      BTSayString(string,0,0,0);

      fflush(stdout);
      if(DosWaitEventSem(sem, -1)) return 1;
      DosResetEventSem(sem,&resetcount);

      if(lapsecs != lapmins*60)
      {
         lapsecs = lapmins*60;
         DosStopTimer(timer);
         if(DosStartTimer(lapsecs * 1000,(HSEM) sem,&timer)) return 1;
      }

      time(&ltime);
      stime = localtime(&ltime);
   }

   DosStopTimer(timer);
   return 0;
}

void help(char *program)
{
   printf("Speaking Clock - Part of the BackTalk project\n"
          "\n"
          "%s [-l <mins>] [-m<mode>] [-d] [-?]\n"
          " -d  disables hour start \n"
          " -l  specifies lap in minutes (default = 15)\n"
          " -m  mode, i=international, c=civil, and o=offset (default = i) \n"
          " -?  this help screen \n",program);
}
